src/main/java/jenkins/plugins/slack/webhook/WebhookEndpoint.java

```diff
@@ -0,0 +1,250 @@
+import hudson.model.UnprotectedRootAction;
+
+import hudson.security.ACL;
+
+import javax.servlet.ServletException;
+
+import java.io.IOException;
+
+import java.util.List;
+import java.util.ArrayList;
+
+import java.util.logging.Logger;
+
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+import jenkins.plugins.slack.webhook.model.JsonResponse;
+import jenkins.plugins.slack.webhook.model.SlackPostData;
+import jenkins.plugins.slack.webhook.model.SlackTextMessage;
+import jenkins.plugins.slack.webhook.model.SlackWebhookCause;
+
+import jenkins.plugins.slack.webhook.exception.CommandRouterException;
+import jenkins.plugins.slack.webhook.exception.RouteNotFoundException;
+
+
+
+
+@Extension
+public class WebhookEndpoint implements UnprotectedRootAction {
+
+    private String slackUser;
+
+    private GlobalConfig globalConfig;
+
+    private static final Logger LOGGER =
+        Logger.getLogger(WebhookEndpoint.class.getName());
+
+    public WebhookEndpoint() {
+        globalConfig = GlobalConfiguration.all().get(GlobalConfig.class);
+    }
+
+    @Override
+    public String getUrlName() {
+        String url = globalConfig.getSlackOutgoingWebhookURL();
+        if (url == null || url.equals(""))
+            return null;
+
+        return "/"+url;
+    }
+
+    @RequirePOST
+    public HttpResponse doIndex(StaplerRequest req) throws IOException,
+        ServletException {
+
+        if (globalConfig.getSlackOutgoingWebhookToken() == null ||
+            globalConfig.getSlackOutgoingWebhookToken().equals("")) {
+            return new JsonResponse(new SlackTextMessage("Slack token not set"),
+                StaplerResponse.SC_OK); 
+        }
+
+        SlackPostData data = new SlackPostData();
+        req.bindParameters(data);
+
+        if (!globalConfig.getSlackOutgoingWebhookToken().equals(data.getToken()))
+            return new JsonResponse(new SlackTextMessage("Invalid Slack token"),
+                StaplerResponse.SC_OK); 
+    
+        String commandText = data.getText();
+        if (commandText == null || commandText.isEmpty())
+            return new JsonResponse(new SlackTextMessage("Invalid command, text field required"),
+                StaplerResponse.SC_OK);
+
+        String triggerWord = data.getTrigger_word();
+        if (triggerWord == null || triggerWord.isEmpty())
+            return new JsonResponse(new SlackTextMessage("Invalid command, trigger_word field required"),
+                StaplerResponse.SC_OK);
+
+        //
+        // TODO: figure out better way of injecting slack user data into handler
+        //
+        this.slackUser = data.getUser_name();
+
+        CommandRouter<SlackTextMessage> router =
+            new CommandRouter<SlackTextMessage>();
+
+        try {
+            router.addRoute("^"+triggerWord+" list projects",
```

Comment:
I was able to perform a DOS attack on the endpoint by using an evil regex as the trigger word. I needed to know the token beforehand so that I could call the endpoint via curl, but it does concern me that a smarter or more determined attacker could do this directly from Slack, which would mean they wouldn't need to have the token. They would need to have Slack access for the relevant team, but there is the disgruntled employee scenario. I don't see how they could do it without the token, but I can't prove it and I wouldn't be surprised if I've missed something.

This is my attack, which ran the CPU hard for 17 mins until I eventually just killed it.

```
$ curl -d 'token=asdasdassdfsdfsdfsdf&trigger_word=(a%2b)%2b&text=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa! list' http://localhost:8080/jenkins/webhook/
```

So it looks like this would be hard to repeat in the wild, but I don't know what uninformed admins will get up to, and they'll have opened a firewall port for this functionality to work (or won't be behind one). I think there is an easy way to totally avoid the potential problem though.

* Return an error if commandText does not begin with the trigger word (check with string match, not a regex)
* Remove the trigger word and whitespace from the beginning of commandText
* Pass that string into your router for matching
* Routes obviously should not include the trigger word in their regexes

That way, user input is never part of the regular expression, which greatly reduces the possibility of abuse. It also has the added advantage that you should be able to build your router once and just keep reusing it, and this section of code should be much more concise.